### ISSUE TITLE
[MASSEMBLY-651] REOPEN - OutOfMemory when running maven-assembly-plugin

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-assembly-plugin#811]
**[Eric Tschetter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=echeddar)** opened **[MASSEMBLY-651](https://issues.apache.org/jira/browse/MASSEMBLY-651?redirect=false)** and commented

I'm experiencing OutOfMemory error when maven build try to complete maven-assembly-plugin job.

Issue reproducible on Linux only.

The project I'm trying to build is [webcam-capture](https://github.com/sarxos/webcam-capture).

Command:

```
$ mvn clean package -P assembly -X
```

It hangs always in the same place:

```
[DEBUG] Adding file-set in: /home/sarxos/workspace-j2se/webcam-capture/webcam-capture/./src/example/java to archive location: examples/
```

In most cases process hangs and I'm getting OutOfMemory. Once I get this exception in the same time, before process started to allocate huge amount of memory OutOfMemory. OldGen is exhausted. Increasing memory to be allocated does not have any effect.

Not sure if exception below is cause or effect.

```
[DEBUG] Adding file-set in: /home/sarxos/workspace-j2se/webcam-capture/webcam-capture/./src/example/java to archive location: examples/
[DEBUG] Cannot find ArtifactResolver with hint: project-cache-aware
org.codehaus.plexus.component.repository.exception.ComponentLookupException: java.util.NoSuchElementException
      role: org.apache.maven.artifact.resolver.ArtifactResolver
  roleHint: project-cache-aware
	at org.codehaus.plexus.DefaultPlexusContainer.lookup(DefaultPlexusContainer.java:257)
	at org.codehaus.plexus.DefaultPlexusContainer.lookup(DefaultPlexusContainer.java:233)
	at org.apache.maven.shared.repository.DefaultRepositoryAssembler.contextualize(DefaultRepositoryAssembler.java:721)
	at org.sonatype.guice.plexus.lifecycles.PlexusLifecycleManager.contextualize(PlexusLifecycleManager.java:317)
	at org.sonatype.guice.plexus.lifecycles.PlexusLifecycleManager.manageLifecycle(PlexusLifecycleManager.java:292)
	at org.sonatype.guice.plexus.lifecycles.PlexusLifecycleManager.onProvision(PlexusLifecycleManager.java:148)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:108)
	at com.google.inject.internal.ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:55)
	at com.google.inject.internal.ProviderInternalFactory.circularGet(ProviderInternalFactory.java:68)
	at com.google.inject.internal.InternalFactoryToInitializableAdapter.get(InternalFactoryToInitializableAdapter.java:45)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1018)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.Scopes$1$1.get(Scopes.java:59)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:41)
	at com.google.inject.internal.InjectorImpl$3$1.call(InjectorImpl.java:965)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1011)
	at com.google.inject.internal.InjectorImpl$3.get(InjectorImpl.java:961)
	at org.sonatype.guice.bean.locators.LazyBeanEntry.getValue(LazyBeanEntry.java:83)
	at org.sonatype.guice.plexus.locators.LazyPlexusBean.getValue(LazyPlexusBean.java:49)
	at org.sonatype.guice.bean.locators.EntryListAdapter$ValueIterator.next(EntryListAdapter.java:112)
	at org.apache.maven.plugin.assembly.archive.DefaultAssemblyArchiver.createArchive(DefaultAssemblyArchiver.java:181)
	at org.apache.maven.plugin.assembly.mojos.AbstractAssemblyMojo.execute(AbstractAssemblyMojo.java:436)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:616)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)
Caused by: java.util.NoSuchElementException
	at java.util.Collections$EmptyIterator.next(Collections.java:2958)
	at org.codehaus.plexus.DefaultPlexusContainer.lookup(DefaultPlexusContainer.java:253)
	... 43 more
```

---

**Affects:** 2.4

**Attachments:**
- [log.txt](https://issues.apache.org/jira/secure/attachment/12714364/log.txt) (_265.29 kB_)

**Issue Links:**
- [MASSEMBLY-646](https://issues.apache.org/jira/browse/MASSEMBLY-646) OutOfMemory when running maven-assembly-plugin
 (_**"duplicates"**_)

**Comment by jira-importer** [comment]
**[Eric Tschetter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=echeddar)** commented

I'm opening a new issue as a clone of this old issue because it is definitely still a problem.  I tried to re-open the old issue, but I think I'm not special enough to do that.

I'm trying to build Druid (http://www.github.com/metamx/druid) and getting the same thing the original reporter did.  I've increased my memory settings up to 4G still have the issue, should a build of a maybe 60MB artifact really take >4GB of memory?

You can reproduce with

```
git clone git@github.com:metamx/druid.git
cd druid
git checkout 12fe145cc87a145157f2e6851226d26ec5b955a7
mvn package
```

Please let me know what I'm doing wrong in configuring this (and make the assembly plugin provide a meaningful error message to indicate what is happening).

I'd like to start using the assembly plugin to build assemblies of my project, but this is blocking me from doing that.

edit: simplify reproduction steps

**Comment by jira-importer** [comment]
**[Eric Tschetter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=echeddar)** commented

Also, I'm running into this issue on mac OSX:

```
Apache Maven 3.0.3 (r1075438; 2011-02-28 11:31:09-0600)
Maven home: /usr/share/maven
Java version: 1.7.0_17, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.7.0_17.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.7.5", arch: "x86_64", family: "mac"
```

edit: add noformat tags

**Comment by jira-importer** [comment]
**[Eric Tschetter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=echeddar)** commented

In the hopes of helping with the debugging effort, we went and took a bunch of thread dumps over and over again while it was building up its memory usage.  It appears to just sit in the same code and accumulate memory.  Here's a trace of where:

```
"main" prio=5 tid=7f95fd801000 nid=0x10ded8000 runnable [10ded5000]
   java.lang.Thread.State: RUNNABLE
	at java.io.UnixFileSystem.list(Native Method)
	at java.io.File.list(File.java:973)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:410)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:521)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:521)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:521)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:521)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:521)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:521)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:521)
	at org.codehaus.plexus.util.DirectoryScanner.scandir(DirectoryScanner.java:521)
	at org.codehaus.plexus.util.DirectoryScanner.scan(DirectoryScanner.java:346)
	at org.codehaus.plexus.components.io.resources.PlexusIoFileResourceCollection.getResources(PlexusIoFileResourceCollection.java:187)
	at org.codehaus.plexus.archiver.AbstractArchiver$1.hasNext(AbstractArchiver.java:468)
	at org.apache.maven.plugin.assembly.filter.ComponentsXmlArchiverFileFilter.finalizeArchiveCreation(ComponentsXmlArchiverFileFilter.java:166)
	at org.codehaus.plexus.archiver.AbstractArchiver.runArchiveFinalizers(AbstractArchiver.java:884)
	at org.codehaus.plexus.archiver.AbstractArchiver.createArchive(AbstractArchiver.java:908)
	at org.apache.maven.plugin.assembly.archive.archiver.AssemblyProxyArchiver.createArchive(AssemblyProxyArchiver.java:512)
	at org.apache.maven.plugin.assembly.archive.DefaultAssemblyArchiver.createArchive(DefaultAssemblyArchiver.java:186)
	at org.apache.maven.plugin.assembly.mojos.AbstractAssemblyMojo.execute(AbstractAssemblyMojo.java:436)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:319)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)
```

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

`@Eric`; your problem is "simple enough". Your assembly descriptor contains a

    <fileSet>
        <directory>/</directory>
        <includes>
            <include>LICENSE</include>
        </includes>
    </fileSet>

Which basically has directoryscanner scanning your complete file system from the root down.  You may not have sufficient ram to do this.

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

I'm wondering, did you find this sample somewhere ?

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

I've made the plugin fail if you attempt to add the entire root file system in r1478985

**Comment by jira-importer** [comment]
**[Eric Tschetter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=echeddar)** commented

LOL, ok, I thought that root was the root of my project, not my system.  Great to know what I did wrong.

I don't remember if I saw that in a sample or if I created it in my own stupidity, but is there a way to make it not recursively search through directories?  I'm certain there are points where I'd like to request that it include the whole recursive structure as well as points where I'd rather it not bother looking through whatever structure happens to exist and just look at the directory specified.
